src/messages/embed/Embed.ts

```diff
@@ -0,0 +1,299 @@
+import ow from 'ow';
+import {
+	authorNamePredicate,
+	colorPredicate,
+	descriptionPredicate,
+	embedFieldsArrayPredicate,
+	fieldInlinePredicate,
+	fieldNamePredicate,
+	fieldValuePredicate,
+	footerTextPredicate,
+	timestampPredicate,
+	titlePredicate,
+	urlPredicate,
+	validateFieldLength,
+} from './Assertions';
+
+/**
+ * Represents an embed in a message (image/video preview, rich embed, etc.)
+ */
+export class Embed implements APIEmbed {
+	/**
+	 * An array of fields of this embed.
+	 */
+	public fields: APIEmbedField[];
+
+	/**
+	 * The embed title.
+	 */
+	public title?: string;
+
+	/**
+	 * The embed description.
+	 */
+	public description?: string;
+
+	/**
+	 * The embed url.
+	 */
+	public url?: string;
+
+	/**
+	 * The embed color.
+	 */
+	public color?: number;
+
+	/**
+	 * The timestamp of the embed in the ISO format.
+	 */
+	public timestamp?: string;
+
+	/**
+	 * The embed thumbnail data.
+	 */
+	public thumbnail?: APIEmbedThumbnail;
+
+	/**
+	 * The embed image data.
+	 */
+	public image?: APIEmbedImage;
+
+	/**
+	 * Received video data.
+	 */
+	public video?: APIEmbedVideo;
+
+	/**
+	 * The embed author data.
+	 */
+	public author?: APIEmbedAuthor;
+
+	/**
+	 * Received data about the embed provider.
+	 */
+	public provider?: APIEmbedProvider;
+
+	/**
+	 * The embed footer data.
+	 */
+	public footer?: APIEmbedFooter;
+
+	public constructor(data: APIEmbed = {}) {
+		this.title = data.title;
+		this.description = data.description;
+		this.url = data.url;
+		this.color = data.color;
+		this.thumbnail = data.thumbnail;
+		this.image = data.image;
+		this.video = data.video;
+		this.author = data.author;
+		this.provider = data.provider;
+		this.footer = data.footer;
+		this.fields = data.fields ?? [];
+
+		if (data.timestamp) this.timestamp = new Date(data.timestamp).toISOString();
+	}
+
+	/**
+	 * The accumulated length for the embed title, description, fields, footer text, and author name.
+	 */
+	public get length(): number {
+		return (
+			(this.title?.length ?? 0) +
+			(this.description?.length ?? 0) +
+			(this.fields.length >= 1
+				? this.fields.reduce((prev, curr) => prev + curr.name.length + curr.value.length, 0)
+				: 0) +
+			(this.footer?.text.length ?? 0) +
+			(this.author?.name?.length ?? 0)
+		);
+	}
+
+	/**
+	 * Adds a field to the embed (max 25).
+	 * @param name The name of this field.
+	 * @param value The value of this field.
+	 * @param inline If this field will be displayed inline.
+	 */
+	public addField(name: string, value: string, inline = false): this {
+		return this.addFields({ name, value, inline });
+	}
+
+	/**
+	 * Adds fields to the embed (max 25).
+	 * @param fields The fields to add.
+	 */
+	public addFields(...fields: APIEmbedField[]): this {
+		// Data assertions
+		ow(fields, embedFieldsArrayPredicate);
+
+		// Ensure adding these fields won't exceed the 25 field limit
+		validateFieldLength(this.fields, fields.length);
+
+		this.fields.push(...Embed.normalizeFields(...fields));
```

Comment:
Same as above, non-normalized fields (e.g. `addFields([{ ... }, { ... }, { ... }])`) will cause wrong checks, in the aforementioned example, I'm adding 3 fields inside an array, so when `validateFieldLength` is called, it'll use the non-normalized fields length (`1`) instead of the flattened ones (`3`).